avalanche/training/plugins/generative_replay.py

```diff
@@ -0,0 +1,153 @@
+        will be set equal to the strategy's batch size.
+    :param batch_size_mem: the size of the memory batch. If
+        `task_balanced_dataloader` is set to True, it must be greater than or
+        equal to the number of tasks. If its value is set to `None`
+        (the default value), it will be automatically set equal to the
+        data batch size.
+    :param task_balanced_dataloader: if True, buffer data loaders will be
+            task-balanced, otherwise it will create a single dataloader for the
+            buffer samples.
+    :param untrained_solver: if True we assume this is the beginning of 
+        a continual learning task and add replay data only from the second 
+        experience onwards, otherwise we sample and add generative replay data
+        before training the first experience. Default to True.
+    """
+
+    def __init__(self, generator=None, mem_size: int = 200, 
+                 batch_size: int = None,
+                 batch_size_mem: int = None,
+                 task_balanced_dataloader: bool = False,
+                 untrained_solver: bool = True):
+        '''
+        Init.
+        '''
+        super().__init__()
+        self.mem_size = mem_size
+        self.batch_size = batch_size
+        self.batch_size_mem = batch_size_mem
+        self.task_balanced_dataloader = task_balanced_dataloader
+        self.generator_strategy = generator
+        if self.generator_strategy:
+            self.generator = generator.model
+        else: 
+            self.generator = None
+        self.untrained_solver = untrained_solver
+        self.model_is_generator = False
+
+    def before_training(self, strategy, *args, **kwargs):
+        """Checks whether we are using a user defined external generator 
+        or we use the strategy's model as the generator. 
+        If the generator is None after initialization 
+        we assume that strategy.model is the generator."""
+        if not self.generator_strategy:
+            self.generator_strategy = strategy
+            self.generator = strategy.model
+            self.model_is_generator = True
+
+    def before_training_exp(self, strategy: "SupervisedTemplate",
+                            num_workers: int = 0, shuffle: bool = True,
+                            **kwargs):
+        """
+        ReplayDataloader to build batches containing examples from both, 
+        data sampled from the generator and the training dataset.
+        """
+
+        if self.untrained_solver:
+            # The solver needs to be trained before labelling generated data and
+            # the generator needs to be trained before we can sample.
+            self.untrained_solver = False
+            return
+
+        # Sample data from generator
+        memory = self.generator.generate(
+            len(strategy.adapted_dataset) *
+            (strategy.experience.current_experience)).to(strategy.device)
```

Comment:
Good question! There are two approaches (as mentioned [here](https://www.nature.com/articles/s41467-020-17866-2#Sec19)): either we generate all data needed for the experience beforehand and then start the training, or we save a copy of the "old" generator and generate our data on demand. I chose the former approach also because this way I could stay in line with the existing Replay strategy, where we update the strategy.data_loader once before each experience. 

Do you you think the other method would be more efficient? How would the implementation of an additional dataloader roughly look like? Since the memory variable is only used temporarily and it is then passed to the ReplayDataloader in the next step, I was hoping that we are indeed already making use of the dataloader properties and do not clog RAM too much (but I am not entirely sure about this).

Reply:
Both are reasonable solutions. If you want to use this one, you should:
- batch the results of the Generator. In general you want minibatch size < data size, because the entire dataset may be too large to generate in one step.
- for the same reasons, unless the dataset is small, it's better to keep it in cpu and move only the minibatch to the gpu when needed.

If you don't do this you will get an out-of-memory error.

Reply:
I see. How would the first point tie in with the ReplayDataLoader? I could generate the data one mini-batch at a time to make it more resistant for larger datasets, but then I would still have to obtain an AvalancheDataset (ie. to concatenate them again?). This is because ReplayDataLoader expects an AvalancheDataset and not some kind of Dataloader, if I understand it correctly?

Reply:
> This is because ReplayDataLoader expects an AvalancheDataset and not some kind of Dataloader, if I understand it correctly?

Exactly. I would make another dataloader similar to ReplayDataLoader that accepts a dataset (the current data), and an iterator (data generator).

Reply:
I definitely like the idea of using a Python Generator. But that would then correspond to the second of the two approaches I mentioned above, as when yielding the next batch we need access to the old version of the generator. I was trying to avoid having to store an additional model as well as trying to stick to the existing ReplayDataLoader as I understood it to be designed for any "rehearsal/replay strategies". 

I am aiming to update the pull request by the beginning of next week.

Reply:
Hi! I just merged my changes ([bba78b9](https://github.com/ContinualAI/avalanche/pull/931/commits/bba78b919f2b6fff66c0b69147c5e914fc9ab9bc)). As mentioned in my last comment I now opted for the "on-demand" replay data generation, where we store the old generator+model and create replay data before each training iteration. This way I got around using the whole ReplayDataLoader syntax and instead just extend the current mini-batch with newly generated replay data.

Accuracy on splitMNIST stayed the same.

Reply:
Good point. I have actually implemented an alternative [solution](https://github.com/travela/avalanche/tree/weighted_loss) where we use a weighted loss to gradually reduce the importance of a new experience/class as the total number grows. As mentioned below, this solution yielded a lower accuracy (in my particular example). However not using any of the two methods yielded much lower accuracy, suggesting that at least one of the two is necessary. 

When implementing the weighted loss I had to overwrite the criterions of both, the model and the generator, which maybe makes it less intuitive, but I guess proper documentation will take care of that.

I was thinking of adding the weighted loss as an option for the user. Maybe I could make it the default option and offer the increasing minibatch size as an alternative (for simple cases like MNIST)?

Reply:
Makes sense. Consider that the simple replay in Avalanche does not change the batch size, so it would be better to have consistent behavior here.